packages/web-media-box/src/dash-parser/parse.ts

```diff
@@ -1,19 +1,109 @@
-import { ParsedManifest } from "./types/parsedManifest";
-import { testString } from "./examples/mpd";
-
-/**
- * Parses a MPD manifest file.
- *
- * @param playlist The URL of the mpd manifest to be parsed.
- */
-export default function parse(playlist: string): ParsedManifest {
-  const parsedManifest: ParsedManifest = {
-    segments: [],
-    custom: {}
-  };
-
-  // TODO: implement parsing.
-  var doc = new DOMParser().parseFromString(testString, 'text/xml');
-
-  return parsedManifest;
+import type { ParsedManifest } from "./types/parsedManifest";
+import { DebugCallback, ParserOptions, WarnCallback } from "./types/parserOptions";
+
+import {
+  ADAPTIONSET,
+  MPD
+} from "./consts/tags";
+import {
+  AdaptionSetProcessor,
+  MPDProcessor
+} from "./nodes/nodesWithAttributes";
+import {
+  MPDAttributes,
+  AdaptationSetAttributes,
+  Attribute
+} from "./defaults/ElementAttributes";
+import { noop } from "../utils/fn";
+import { NodeProcessor } from "./nodes/nodeProcessor";
+
+class Parser {
+  // We will keep the state of the iteration here, so we can use information
+  // even after we have iterated over a node.
+
+  private currentNode: Element;
+  private readonly warnCallback: WarnCallback;
+  private readonly debugCallback: DebugCallback;
+  private readonly nodeProcessorMap: Record<string, NodeProcessor>;
+  private readonly nodeAttributesMap: Record<string, Array<Attribute>>;
+
+  protected readonly parsedManifest: ParsedManifest;
+
+  public constructor(mpd: string, options: ParserOptions) {
+    this.warnCallback = options.warnCallback || noop;
+    this.debugCallback = options.debugCallback || noop;
+    this.currentNode = {} as Element;
+
+    this.parsedManifest = {
+      uri: '',
+      representations: [],
+    };
+
+    this.nodeProcessorMap = {
+      [MPD]: MPDProcessor,
+      [ADAPTIONSET]: AdaptionSetProcessor,
+    };
+
+    this.nodeAttributesMap = {
+      [MPD]: MPDAttributes,
+      [ADAPTIONSET]: AdaptationSetAttributes,
+    };
```

Comment:
you probably dont need 2 maps here, since you can 

```ts
   this.nodeProcessorMap = {
      [MPD]: new MPDProcessor(),
      [ADAPTIONSET]: new AdaptionSetProcessor(),
    };
    
    Attributes maps are constants and can be a static member of each particular processor

```

Reply:
You do not need to provide the current node to the constructor of each processor. You can expect it to be an argument of the process method alongside parsed manifest object and any other shared object (if needed)